content/blog/2019-10-24-superblock-scheduling.md

```diff
@@ -0,0 +1,251 @@
+a bundle.
+
+```
+b1:
+  v1: int = add v2 v3
+  br v7 b2 b3
+b2:
+  v4: int = add v2 v0 // not live out of b2
+  ...
+b3:
+  ...
+```
+
+The core idea with superblock scheduling is finding "traces" of frequently
+code by predicting which branch a program is going to take and building
+a fast program path with dense bundles. In case the branch prediction is
+wrong, the compiler adds abort labels to exit the trace.
+
+With superblock scheduling, the code example above can be turned into:
+
+```
+b1:
+  [ v7 : v1: int = add v2 v3, v4: int = add v2 v0 : slow ]
+  br v7 b2 b3
+slow:
+  v1: int = add v2 v3
+  br v7 b2 b3
+b2:
+  v4: int = add v2 v0 // not live out of b2
+  ...
+b3:
+  ...
+```
+
+In the fast case (when `v7` is true), the program will compute both `v1` and
+`v4` in parallel. If `v7` is false, the program switches back to normal
+execution, computing the `v1` and `v4` sequentially. A superblock compiler
+might choose to make various part of the "slow" program paths traces
+themselves, trading off program size for speed.
+
+### List Scheduling
+
+List scheduling is a heuristic-based algorithm that performs compaction
+on a sequence of instructions. It takes a directed acyclic graph (DAG) representing the
+scheduling constraints (such as read-write dependencies) between instructions, and returns a list of bundles.
+
+We start by presenting a high level overview of the algorithm and then we will go into more detail
+about how we build the DAG. First, a heuristic assigns each instruction in the graph a priority.
+Then initialize a queue with all the instructions
+that have no predecessors in the graph. Until we have scheduled all the instructions,
+we do the following in a loop:
+- Make an bundle.
+- Take the instruction from the queue with the highest priority.
+- Add it to the bundle if compatible i.e. has no conflicts with the bundle.
+- Continue taking instructions from the queue until either the queue is empty or the instruction
+is incompatible with the current bundle.
+- For each element that we scheduled, check if their successors now have all their predecessors scheduled and add them to the queue
+
+The algorithm schedules a instruction only after all of its predecessors in the
+DAG have been scheduled. This means that to maintain program correctness,
+the predecessor relation must respect data dependencies.
+
+```
+v1: int = id x;
+x: int = const 10;
+```
+
+In the program above, we cannot reorder the instructions because the second
+instruction writes to `x` while the first reads from it. We can encode this
+constraint in the DAG by making the first instruction a predecessor of the second.
+
+### Generating Superblock
+
+The superblock algorithm generates a sequence of instruction and DAG that are
+used by the list scheduling algorithm to generate a program trace. This means
+that the superblock algorithm has to encode additional dependencies in the DAG
+to allow an unmodified list scheduling algorithm to work.
```

Comment:
This sentence sort of seems to come out of nowhere—you back it up in the next paragraph, but it sounds strange because the second sentence doesn't logically follow from the first (there's no implication between "the superblock algorithm generates…" and "the superblock algorithm has to encode additional dependencies…").

Maybe just quickly say that the DAG has to encode both data and control dependencies, and then get on to your example to explain what that means?